rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,213 @@
+                                        branch/v14 |----------|----------|---------....---->
+                                                   |
+                                                   |
+            14.0.0-cloud.20230712T123633.aabbcc      14.0.0-cloud...  15.0.0-cloud...    15.0.0-cloud...    16.0.0-cloud...
+master -----|-------....---------------------------|------------------|-------....-------|------------------|--------------....---->
+                                                                      |
+                                                                      |
+                                                           branch/v15 |---------|---------|---------...---->
+                                                                      15.0.0    15.0.1    15.0.2
+```
+
+Once a self-hosted release branch is cut, subsequent self-hosted minor/patch
+releases are published without any changes to the current process, but on a
+less frequent cadence than Cloud releases off of master. Changes to the release
+branches will follow the same "develop on master, backport to release branch"
+workflow we're using now.
+
+### Cloud release artifacts
+
+The self-hosted release and promotion process is slow and takes 2-3 hours
+because it builds a full suite of artifacts. Cloud users need most of these
+artifacts as well for the agents they run on their own infrastructure, however
+Cloud control plane needs a very limited subset of all artifacts. For this
+reason, the Cloud release and promotion process will be split in 2 parts:
+
+- The workflow that builds artifacts required for the control plane: amd64
+  binaries and Docker images.
+- The workflow that builds remaining artifacts such as deb/rpm packages and
+  Helm charts.
+
+This will allow us to produce rapid releases that affect only control plane
+services (auth, proxy) while still having the ability to push remaining artifacts
+when needed.
+
+Both workflows will run in Github Actions and will reuse existing bits of the
+full self-hosted release pipeline.
+
+### Cloud promotion process
+
+The Cloud release promotion process will be staged, with a different class of
+tenants being upgraded at each stage:
+
+1. Canary
+2. Internal
+3. Trials
+4. Team & Enterprise
+
+The canary tenants will be a set of precreated clusters with some data in them
+which engineers will use for connecting their agents and testing their features.
+Canary clusters will be updated automatically once a Cloud release is cut.
+
+Before promoting the release further, engineers will test any significant change
+in the cloud staging environment and then explicitly sign off that the change is
+ready for deployment by applying a label (e.g. `cloud/verified`) on the master
+PR corresponding to the change. The release manager then will approve the
+further rollout.
+
+Once the change is rolled out to production tenants, engineers will be expected
```

Comment:
If it is needed to coordinate the rollout of new monitoring to be able to effectively monitor a new feature, how should the rollout of that be coordinated with the rollout of the cloud release? I'd suggest that it should be in place before promotion from Canary to Internal and that should be one of the gates to promotion.

Reply:
I think it depends on the kind of monitoring. For product-related metrics (like, usage) I think it's fine to deploy it together with the feature (in practice, product metrics actually even lag behind feature implementation). For performance metrics, work is being done currently to implement those for various protocols we support so we'll build dashboards we'll be able to watch during the rollout (and setup alerting in future).